apps/emqx/src/emqx_broker_mon.erl

```diff
@@ -178,6 +208,203 @@ update_broker_pool_max_mailbox_size(State) ->
     atomics:put(Atomics, Idx, MailboxSize),
     MailboxSize.
 
+update_top_n_mailbox_procs(ProcsInfo) ->
+    ets:delete_all_objects(?MAILBOX_LEN_TAB),
+    lists:foreach(
+        fun({Pid, QLen, ProcName}) ->
+            ets:insert(?MAILBOX_LEN_TAB, {Pid, ProcName, QLen, make_proc_meta_data(Pid)})
+        end,
+        ProcsInfo
+    ).
+
+make_proc_meta_data(Pid) ->
+    case erlang:process_info(Pid, dictionary) of
+        undefined ->
+            #{};
+        {dictionary, Dict} ->
+            LoggerProcMeta = proplists:get_value('$logger_metadata$', Dict, #{}),
+            emqx_utils_maps:jsonable_map(
+                LoggerProcMeta,
+                fun(K, V) ->
+                    {emqx_utils_conv:bin(K), emqx_utils_conv:bin(V)}
+                end
+            )
+    end.
+
+non_zero_mailbox_procs(TopN) ->
+    [
+        {Pid, QLen, proc_name(Pid)}
+     || {Pid, QLen, _} <- recon:proc_count(message_queue_len, TopN), QLen > 0
```

Comment:
This is extremely costly, because it implies inspecting the runtime state of _the whole system_. And when the system is overloaded, it would be even more taxing: it wiil occupy scheduler for a while, and will likely additionally impair progress because of all the synchronization it needs to do to get the accurate mailbox size.

Reply:
Simply quiring the mailbox size is not *too* costly even if the number of processes is in the million range. Getting the *contents* of the mailbox, on the other hand, can easily OOM the VM.